engine/beef_bump.go

```diff
@@ -2,130 +2,123 @@ package engine
 
 import (
 	"context"
-	"sort"
 
+	"github.com/bitcoin-sv/go-sdk/chainhash"
 	trx "github.com/bitcoin-sv/go-sdk/transaction"
 	"github.com/bitcoin-sv/spv-wallet/engine/spverrors"
 )
 
-func calculateMergedBUMP(txs []*Transaction) (BUMPs, error) {
-	bumps := make(map[uint64][]BUMP)
-	mergedBUMPs := make(BUMPs, 0)
-
-	for _, tx := range txs {
-		if tx.BUMP.BlockHeight == 0 || len(tx.BUMP.Path) == 0 {
-			continue
-		}
+type beefPreparationContext struct {
+	ctx            context.Context
+	store          TransactionGetter
+	transactionMap map[chainhash.Hash]*trx.Transaction
+	visited        map[chainhash.Hash]bool
+	queue          []*trx.Transaction
+	txsForBEEF     []*trx.Transaction
+}
 
-		bumps[tx.BlockHeight] = append(bumps[tx.BlockHeight], tx.BUMP)
+func prepareBEEFFactors(ctx context.Context, tx *Transaction, store TransactionGetter) ([]*trx.Transaction, error) {
+	txsForBEEF, err := initializeRequiredTxsCollection(tx)
+	if err != nil {
+		return nil, err
 	}
 
-	// ensure that BUMPs are sorted by block height and will always be put in beef in the same order
-	mapKeys := make([]uint64, 0, len(bumps))
-	for k := range bumps {
-		mapKeys = append(mapKeys, k)
+	bctx := &beefPreparationContext{
+		ctx:            ctx,
+		store:          store,
+		transactionMap: make(map[chainhash.Hash]*trx.Transaction),
+		visited:        make(map[chainhash.Hash]bool),
+		queue:          []*trx.Transaction{},
+		txsForBEEF:     txsForBEEF,
 	}
-	sort.Slice(mapKeys, func(i, j int) bool { return mapKeys[i] < mapKeys[j] })
 
-	for _, k := range mapKeys {
-		bump, err := CalculateMergedBUMP(bumps[k])
-		if err != nil {
-			return nil, spverrors.Wrapf(err, "failed to calculate merged BUMP")
-		}
-		if bump == nil {
-			continue
+	processedSDKTx := txsForBEEF[0]
+	txID := *processedSDKTx.TxID()
+	bctx.transactionMap[txID] = processedSDKTx
+	bctx.queue = append(bctx.queue, processedSDKTx)
+
+	for len(bctx.queue) > 0 {
+		currentTx := bctx.queue[0]
+		bctx.queue = bctx.queue[1:]
```

Comment:
Can there be a situation when queue has only 1 item inside? Then this line would error

Reply:
This line will not error when in queue we have only one item.
[Pop Front/Shift](https://go.dev/wiki/SliceTricks#pop-frontshift)
`x, a = a[0], a[1:]`
as long as shift value will be lower the slice capacity there are no issues.
So this line:
`len(bctx.queue) > 0 {`
ensures that capacity will be bigger than 0 when we enter loop

Reply:
Ahh I see smart :D